### DISCUSSION TITLE
Single statement returns with out parameters

### DISCUSSION THREAD

**Discussion by BalassaMarton** [dotnet/csharplang#7191] [Language Ideas]
Not sure if this was proposed before, I couldn't find it. When a method has a return type and `out` parameters, it gets verbose to populate them. Consider this made up function that determines if a string needs escaping, and returns the first escaped character index if so:

```cs
    static bool NeedsEscaping(string? input, out int pos)
    {
        if (string.IsNullOrEmpty(input))
        {
            pos = -1;

            return false;
        }

        if (input.StartsWith(' '))
        {
            pos = 0;

            return true;
        }

        for (var i = 0; i < input.Length; i++)
        {
            if (input[i] is '"' or '\'' or '\\')
            {
                pos = i;

                return true;
            }
        }

        if (input.EndsWith(' '))
        {
            pos = input.Length - 1;

            return true;
        }

        pos = -1;

        return false;
    }
```

With the proposed new syntax, return statements can include assignments for the out parameters:

```cs
    static bool NeedsEscaping(string? input, out int pos)
    {
        if (string.IsNullOrEmpty(input))
            return false with pos = -1;

        if (input.StartsWith(' '))
            return true with pos = 0;

        for (var i = 0; i < input.Length; i++)
        {
            if (input[i] is '"' or '\'' or '\\')
                return true with pos = i;
        }

        if (input.EndsWith(' '))
            return true with pos = input.Length - 1;

        return false with pos = -1;
    }
```

When there are multiple out parameters, they can be assigned in a single statement, separating assignments with a comma:

```cs
return true with p1 = value1, p2 = value2, p3 = value3;
```

Which would be lowered into:

```cs
{
    p1 = value1;
    p2 = value2;
    p3 = value3;
    return true;
}
```

**Comment by HaloFour** [comment]
```cs
pos = -1; return false;
```

is already legal and shorter than:

```cs
return false with pos = -1;
```

**Reply by BalassaMarton** [reply]
That's two statements that linters will break into multiple lines.

**Comment by DavidArno** [comment]
A tuple can be used to avoid the need for `out`:
```cs
static (bool needed, int pos) NeedsEscaping(string? input)
{
    if (string.IsNullOrEmpty(input)) return (false, -1);
    if (input.StartsWith(' ')) return (true, 0);
    
    for (var i = 0; i < input.Length; i++)
    {
        if (input[i] is '"' or '\'' or '\\') return (true, i);
    }

    if (input.EndsWith(' ')) return (true, input.Length - 1);

    return (false, -1);
}
```
and that can be matched in much the same way:
```cs
if (NeedsEscaping(someString) is (true, var pos))
{
    // do something with pos...
}
```
Out parameters really aren't needed anymore as we have NRTs, records, tuples and hopefully soon unions that all handle multiple return values far better. Further, Microsoft's own quality analyzer rule, [CA1021: Avoid out parameters](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1021) discourages the use of out parameters.  Therefore the case for extending syntax around out parameters really feels far from compelling to me.

**Reply by BalassaMarton** [reply]
I get it, but this is not helpful when there are legacy API constaints.

**Reply by HaloFour** [reply]
> Further, Microsoft's own quality analyzer rule, [CA1021: Avoid out parameters](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1021) discourages the use of out parameters.

IMO that does not apply to the `bool Try(out X foo)` pattern, which is considered idiomatic both in C# and in the runtime.

**Comment by BalassaMarton** [comment]
Closing it as @DavidArno somewhat explained why this will never happen
> Further, Microsoft's own quality analyzer rule, [CA1021: Avoid out parameters](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1021) discourages the use of out parameters. Therefore the case for extending syntax around out parameters really feels far from compelling to me.
